### PR TITLE
Select a windows reference module

### PR DESCRIPTION
**Pull request by dagwieers**
##### ISSUE TYPE

 - Docs Pull Request

##### COMPONENT NAME

win_service

##### ANSIBLE VERSION

v2.3

##### SUMMARY

As discussed in #20505 we would like to point from the documentation to
a module we choose as the reference module for Windows.

I picked win_service and cleaned it up to the best of my abilities.
I also added check-mode support as an example.

Let the bikeshedding begin !

### CODE REVIEW

**Comment by dagwieers** [comment]
So, let me start off by pointing out that:
- This module does not have any `-type` set for any parameter. (Fixed with ansible/ansible#20124)
- The module does not have any extended documentation, but let us first agree on the module/logic/semantics first and then add the proper code comments.
- This PR will also include documentation updates pointing to this module, as well as providing a list of best practices (including pointing to powershell.ps1 to see what functionality is already provided).

**Comment by mattclay** [comment]
Closing and re-opening to trigger CI.

File: lib/ansible/modules/windows/win_service.ps1 [hunk 1]
```diff
@@ -19,87 +19,80 @@
...
-If ($state) {
-    If ($state -eq "started" -and $svc.Status -ne "Running") {
-        try {
-            Start-Service -Name $svcName -ErrorAction Stop
-        }
-        catch {
+if ($state -eq "started" -and $service.Status -ne "Running") {
+
+    if (-not $check_mode) {
+        Try {
```
Comment (ilpianista) [inline lib/ansible/modules/windows/win_service.ps1]: Uppercase `Try` here

Comment (dagwieers) [reply lib/ansible/modules/windows/win_service.ps1]: Fixed. Thanks !

**Comment by dagwieers** [comment]
Shippable bailing out again:

    ERROR: 500: An error occurred (InstanceLimitExceeded) when calling the RunInstances operation: Your quota allows for 0 more running instance(s). You requested at least 1

**Comment by mattclay** [comment]
@dagwieers It looks like there are legitimate CI failures now:

```
2017-01-26 19:27:52 TASK [win_service : check that enabling the service succeeds for manual startup] ***
2017-01-26 19:27:52 task path: /root/src/github.com/ansible/ansible/test/integration/targets/win_service/tasks/main.yml:81
2017-01-26 19:27:52 fatal: [windows-2016-English-Full-Base]: FAILED! => {
2017-01-26 19:27:52     "assertion": "win_service_manual_start_mode.state == 'stopped'", 
2017-01-26 19:27:52     "changed": false, 
2017-01-26 19:27:52     "evaluated_to": false, 
2017-01-26 19:27:52     "failed": true, 
2017-01-26 19:27:52     "invocation": {
2017-01-26 19:27:52         "module_args": {
2017-01-26 19:27:52             "that": [
2017-01-26 19:27:52                 "win_service_manual_start_mode|changed", 
2017-01-26 19:27:52                 "win_service_manual_start_mode.start_mode == 'manual'", 
2017-01-26 19:27:52                 "win_service_manual_start_mode.state == 'stopped'"
2017-01-26 19:27:52             ]
2017-01-26 19:27:52         }, 
2017-01-26 19:27:52         "module_name": "assert"
2017-01-26 19:27:52     }
2017-01-26 19:27:52 }
```

**Comment by gundalow** [comment]
win_service integration tests are failing, needs_revision

**Comment by dagwieers** [comment]
I am sorry, but I don't see anything indicating the module is not working correctly.
Somehow it feels as if there is a genuine issue with the operating system, or the tests itself.

```
2017-01-26 17:43:22 TASK [win_service : test win_service module with short name] *******************
2017-01-26 17:43:23 fatal: [windows-2012-R2_RTM]: FAILED! => {"changed": false, "failed": true, "msg": "Service 'SSDP Discovery (SSDPSRV)' cannot be started due to the following error: Cannot start service SSDPSRV on computer '.'."}
2017-01-26 17:43:24 fatal: [windows-2008-SP2]: FAILED! => {"changed": false, "failed": true, "msg": "Service 'SSDP Discovery (SSDPSRV)' cannot be started due to the following error: Cannot start service SSDPSRV on computer '.'."}
2017-01-26 17:43:24 fatal: [windows-2008-R2_SP1]: FAILED! => {"changed": false, "failed": true, "msg": "Service 'SSDP Discovery (SSDPSRV)' cannot be started due to the following error: Cannot start service SSDPSRV on computer '.'."}
2017-01-26 17:43:26 fatal: [windows-2012-RTM]: FAILED! => {"changed": false, "failed": true, "msg": "Service 'SSDP Discovery (SSDPSRV)' cannot be started due to the following error: Cannot start service SSDPSRV on computer '.'."}
2017-01-26 17:43:39 ok: [windows-2016-English-Full-Base] => {"changed": false, "display_name": "SSDP Discovery", "name": "SSDPSRV", "start_mode": "manual", "state": "running", "warnings": []}
```

For some reason the service does not start when we expect it to start. But that is unrelated to the changes I have made. The fact that it does work on one of the targets is peculiar as well. (I have some check_mode changes I want to test and this will trigger a rebuild, let's see what it does now)

**Comment by mattclay** [comment]
We've had trouble with the service test before, but the problem has always been intermittent. Here it's failing on all four versions tested, and again after the automatic retry. It doesn't mean the problem is with the module, but if not, the changes at least make the test unusable.

@nitzmahone and I have discussed updating the test to use a minimal purpose-built service for the test, instead of finding an existing one to commandeer for tests. It looks like this PR may force the issue.

**Comment by jborean93** [comment]
@dagwieers @ilpianista I was hoping to get all out changes to win_service in before the next release but it seems like we have 3 different pull requests going on at the same time;

* one to update to newer module standards (this one)
* one to add delayed start to the start_mode (https://github.com/ansible/ansible/pull/20124)
* one to add in more options and get status of service (https://github.com/ansible/ansible/pull/19143)

I am happy to modify my pull request to add in all your changes instead of trying to make 3 different changes ending up in some merge issues but wanted to get your thoughts first. Are you happy for me to do so or would we just want to get a free for all :)

**Comment by dagwieers** [comment]
@jborean93 Thanks for the offer, but this PR is going to be a brainstorm session related to how the standard (reference) windows module should look like. So it's likely that we will discuss and make changes to it the coming weeks.

Please promote your 2 changes independently and I don't mind to fix any conflicts that arise.

**Comment by dagwieers** [comment]
@jborean93 You are right after reviewing #19143 it makes no sense to keep this around. I'll be reviewing your PR.

**Pull request closed without merging**